tests/Doctrine/Tests/ORM/Functional/Ticket/GH10288Test.php

```diff
@@ -0,0 +1,230 @@
+     *
+     * @group GH-10288
+     */
+    public function testEnumDiscriminatorsShouldBeConvertedToString(): void
+    {
+        $boss         = new GH10288BossJoined('John');
+        $boss->bossId = 1;
+        $employee     = new GH10288EmployeeJoined('Bob');
+
+        $boss2         = new GH10288BossSingleTable('John');
+        $boss2->bossId = 1;
+        $employee2     = new GH10288EmployeeSingleTable('Bob');
+
+        $this->_em->persist($boss);
+        $this->_em->persist($boss2);
+        $this->_em->persist($employee);
+        $this->_em->persist($employee2);
+        $this->_em->flush();
+        $this->_em->clear();
+
+        // Using DQL here to make sure that we'll use ObjectHydrator instead of SimpleObjectHydrator
+        $query = $this->_em->createQueryBuilder()
+            ->select('person')
+            ->from(GH10288PersonJoined::class, 'person')
+            ->where('person.name = :name')
+            ->setMaxResults(1)
+            ->getQuery();
+
+        $query->setParameter('name', 'John');
+        self::assertEquals($boss, $query->getOneOrNullResult(AbstractQuery::HYDRATE_OBJECT));
+        self::assertEquals(
+            GH10288People::BOSS,
+            $query->getOneOrNullResult(AbstractQuery::HYDRATE_ARRAY)['discr']
+        );
+
+        $bossId = $boss->id;
+
+        // Using DQL here to make sure that we'll use ObjectHydrator instead of SimpleObjectHydrator
+        $query2 = $this->_em->createQueryBuilder()
+            ->select('person')
+            ->from(GH10288PersonSingleTable::class, 'person')
+            ->where('person.name = :name')
+            ->setMaxResults(1)
+            ->getQuery();
+
+        $query2->setParameter('name', 'Bob');
+        self::assertEquals($employee2, $query2->getOneOrNullResult(AbstractQuery::HYDRATE_OBJECT));
+        self::assertEquals(
+            GH10288People::EMPLOYEE,
+            $query2->getOneOrNullResult(AbstractQuery::HYDRATE_ARRAY)['discr']
+        );
+
+        $this->_em->clear();
+
+        // test SimpleObjectHydrator
+        $bossFetched = $this->_em->find(GH10288PersonJoined::class, $bossId);
+        self::assertEquals($boss, $bossFetched);
+    }
+}
+
+class GH10288PeopleType extends StringType
+{
+    public const NAME = 'GH10288people';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function convertToDatabaseValue($value, AbstractPlatform $platform)
+    {
+        if (! $value instanceof GH10288People) {
+            $value = GH10288People::from($value);
+        }
+
+        return $value->value;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function convertToPHPValue($value, AbstractPlatform $platform)
+    {
+        return GH10288People::from($value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
+    {
+        return self::NAME;
+    }
+}
+
+/**
+ * @Entity
+ * @InheritanceType("JOINED")
+ * @DiscriminatorColumn(name="discr", type="GH10288people")
```

Comment:
Why do I still need a custom type for my enum?

Reply:
`DiscriminatorColumn` is not a `Column`, you cannot specify `enumType`. So the only way to emulate the case where discriminator column gets a PHP enum assigned to it during hydration is to use a custom type which returns `BackedEnum` for its `convertToPhpValue`

Reply:
> `DiscriminatorColumn` is not a `Column`, you cannot specify `enumType`

I know, but I would kinda expect that a PR that bears the title "Allow enum discriminator columns" would change exactly that. 🤔 

As long as this custom type is needed, the whole feature feels like a hack to me.

Reply:
but there is absolutely no point to change this. discriminator column is special, it is not accessible from user-end code. The column is defined only at the top of the Entity, and there is no PHP field that can be accessed when working with such entity.
The issue arises, when you try to use custom type, which works with enums. This is because PHP does not support `__toString()` overload on enums. Objects are supported as discriminator columns exactly by leveraging `__toString()` override. Since this is not possible with enums, the only way is to check with `instanceof` and use enum's value.
There is absolutely no reason not use to allow adding `enumType` into the discriminator column. There is reason however to allow custom types using enums, for the exact same reason you want to allow objects to be used as discriminator columns (implemented with https://github.com/doctrine/orm/pull/6141)

Reply:
> discriminator column is special, it is not accessible from user-end code

And why would I need a custom type for an inaccessible column?

> The column is defined only at the top of the Entity, and there is no PHP field that can be accessed when working with such entity.

You can still select and hydrate that column, like you did in your test, right?

Reply:
> And why would I need a custom type for an inaccessible column?

Well e.g. in my case, I use custom type that creates native MySQL ENUM. When converting to PHP, it converts to PHP enum value thats specified by the given type. So if I want to have in mysql DB the field as ENUM, I have to use a custom type.

> You can still select and hydrate that column, like you did in your test, right?

ok, let me try to implement this